File: lib/product.dart
class Product {
  int? id;
  String? title;
  String? description;
  DateTime? creationDate;
  int? creatorUserId;
  bool? isDeleted;
  String? imageUrl;

  Product(this.id, this.title, this.description, this.creationDate, this.creatorUserId, this.isDeleted, this.imageUrl);

  Product.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    title = json['title'];
    description = json['description'];
    creationDate = json['creationDate'];
    creatorUserId = json['creatorUserId'];
    isDeleted = json['isDeleted'];
    imageUrl = json['imageUrl'];
  }

  factory Product.fromMap(Map<String, dynamic> json) {
    return Product(
      json['id'],
      json['title'],
      json['description'],
      json['creationDate'],
      json['creatorUserId'],
      json['isDeleted'],
      json['imageUrl'],
    );
  }
}

File: crudApp/src/main/java/com/example/crudApp/model/Product.java
package com.example.crudApp.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import lombok.Getter;
import lombok.Setter;
import org.springframework.security.core.context.SecurityContextHolder;

import java.time.LocalDateTime;
import java.util.Set;

@Entity
@Table
@Getter
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;
    @NotBlank(message = "Product title must not be empty.")
    @Setter
    private String title;
    @NotBlank(message = "Product description must not be empty.")
    @Setter
    private String description;
    private LocalDateTime creationDate;
    private String creatorUserId;
    private boolean isDeleted;
    @Setter
    private String imageUrl;
    @OneToMany(mappedBy = "product")
    @Setter
    private Set<Comment> comments;
    @ManyToMany(mappedBy = "products", fetch = FetchType.EAGER)
    @Setter
    private Set<Category> categories;

    public boolean isDeleted() {
        return isDeleted;
    }

    public Product() {}
    public Product(String title, String description, String imageUrl, Set<Category> categories) {
        this.title = title;
        this.description = description;
        this.imageUrl = imageUrl;
        this.categories = categories;
    }

    @PrePersist
    void PrePersist() {
        creationDate = LocalDateTime.now();
        isDeleted = false;
        for (Category category : categories) {
            category.addProduct(this);
        }
        //creatorUserId = SecurityContextHolder.getContext().getAuthentication().getName();
    }

    public void addComment(Comment comment) {
        comments.add(comment);
    }

    public void delete () {
        isDeleted = true;
        for (Comment comment : comments) {
            comment.remove();
        }
    }

    public void update (Product source) {
        this.title = source.title;
        this.description = source.description;
        this.imageUrl = source.imageUrl;
        this.categories.clear();
        this.categories.addAll(source.categories);
    }
}